### ISSUE TITLE
Discrepancy between CPU and GPU experiments

### ISSUE THREAD

**Issue by timmoon10** [LLNL/lbann#100]
Results from AlexNet experiments (10 classes of ILSVRC2012, minibatch 256, Adam) on Surface (pbatch partition, 8 nodes, 16 MPI ranks):

### CPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy  | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | 57.2s | 3.13 | 9.95% | 48.87% | 10% | 50% |
| 2 | 51.7s | 2.32 | 12.02% | 54.38% | 16.2% | 67% |
| 5 | 53.1s | 1.59 | 41.19% | 90.50% | 39% | 90.4% |
| 10 | 51.4s | 1.19 | 56.16% | 96.03% | 52.8% | 93.2% |

### GPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy  | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | 14.8s | 3.16 | 10.35% | 49.38% | 15.2% | 63.2% |
| 2 | 14.1s | 2.12 | 20.61% | 71.32% | 20% | 81.2% |
| 5 | 14.4s | 1.50 | 44.14% | 92.30% | 43.8% | 91.6% |
| 10 | 14.4s | 1.12 | 59.06% | 96.50% | 54% | 94% |

**Comment by naoyam** [comment]
Is this float? Have you tried double?

**Comment by timmoon10** [comment]
Results with `double`:

### CPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy  | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | 74.9s | 3.34 | 9.62% | 48.967% | 10% | 50.2% |
| 2 | 75.5s | 2.24 | 13.91% | 62.82% | 17% | 76.2% |
| 5 | 75.0s | 1.80 | 33.48% | 86.58% | 35.6% | 89.2% |
| 10 | 74.7s | 1.27 | 52.2% | 94.76% | 45.6% | 90% |

### GPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy  | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | 34.1s | 3.18 | 9.96% | 49.59% | 10.6% | 47.8% |
| 2 | 27.4s | 2.23 | 16.78% | 59.86% | 22% | 76.2% |
| 5 | 27.4s | 1.59 | 40.43% | 90.86% | 40.2% | 90.6% |
| 10 | 27.6s | 1.19 | 55.79% | 95.59% | 50% | 94% |

**Comment by bvanessen** [comment]
Do we have a consistent implementation of all of the layers.  I am specifically thinking about batchnorm, but I don’t think that is used in our Alexnet implementation.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by timmoon10** [comment]
We should have consistent implementations. There is probably a bug lurking somewhere.

**Comment by naoyam** [comment]
What is the expected range of difference between CPU and GPU? I can imagine GPU may have better accuracy as FMA rather than MUL+ADD would be more often used on GPU than on CPU.

**Comment by timmoon10** [comment]
The CPU code rearranges data with an im2col and then performs a GEMM. cuDNN supports several algorithms, some of which are similar to ours and others which use FFTs or Winograd convolution.

**Comment by bvanessen** [comment]
Can we force cuDNN to use im2col for comparison purposes?


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Aug 11, 2017, at 4:11 PM, Tim Moon <notifications@github.com<mailto:notifications@github.com>> wrote:


The CPU code rearranges data with an im2col and then performs a GEMM. cuDNN supports several algorithms, some of which are similar to ours and others which use FFTs or Winograd convolution.

**Comment by bvanessen** [comment]
Does Nikoli's fix with the elemental views change these results?


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Aug 11, 2017, at 6:00 PM, Brian Van Essen <notifications@github.com<mailto:notifications@github.com>> wrote:

Can we force cuDNN to use im2col for comparison purposes?


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov><mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Aug 11, 2017, at 4:11 PM, Tim Moon <notifications@github.com<mailto:notifications@github.com><mailto:notifications@github.com>> wrote:


The CPU code rearranges data with an im2col and then performs a GEMM. cuDNN supports several algorithms, some of which are similar to ours and others which use FFTs or Winograd convolution.

**Comment by ndryden** [comment]
My fixes for the views did not affect accuracy, just performance.

**Comment by timmoon10** [comment]
I've fixed a bug in the CPU LRN layer and forced cuDNN to use the im2col GEMM algorithm (the back prop does not construct the im2col matrix explicitly).

### Experiment parameters
- AlexNet
- 10 classes of ILSVRC2012
- Minibatch size 256
- Adam optimizer
- `double` datatype
- Run on Surface cluster, pbatch partition
- 8 nodes, 2 tasks/node
- Dataset cached locally

### CPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy  | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | 74.8s | 3.33 | 9.81% | 49.05% | 15% | 53.4% |
| 2 | 74.3s | 2.21 | 16.08% | 66.29% | 16.2% | 79% |
| 5 | 74.4s | 1.62 | 40.33% | 90.87% | 41% | 90.8% |
| 10 | 73.7s | 1.29 | 52.65% | 95.03% | 50.8% | 93.2% |

### GPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy  | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | 158.7s | 3.18 | 10.17% | 49.91% | 15.8% | 52.8% |
| 2 | 159.6s | 2.23 | 17.10% | 59.15% | 18% | 65.6% |
| 5 | 158.7s | 1.55 | 42.87% | 91.22% | 42.8% | 92.6% |
| 10 | 159.4s | 1.15 | 57.09% | 96.13% | 53.2% | 95.4% |

**Comment by ndryden** [comment]
Can we do gradient checking to possibly figure out whether there are further bugs in the CPU code?

**Comment by timmoon10** [comment]
Gradient check experiments indicate that the CPU and GPU code are both correct when compiled with `float`. When compiled with `double`, the CPU code is correct and the GPU code has error consistent with single precision. I suspect that cuDNN's convolution operation contains an internal single-precision operation.

**Comment by bvanessen** [comment]
Great.  Thanks for doing that.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by timmoon10** [comment]
It turns out that our CPU convolution layer is performing cross-correlation, not convolution. This is what Caffe and Tensorflow do, so I've modified our GPU convolution layer in bbdf90f8649f9e6a2a7bea32f7d928d24cf40c20. Here are some new experiments:

### Parameters
- AlexNet without image crops, 256 mini-batch, momentum SGD
- Trained on first 10 classes of ILSVRC2012, cached in /tmp
- Release build, `float` data type
- Run on Surface cluster, gpgpu partition: 8 nodes, 2 procs/node

### CPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|---|---|---|---|---|---|---|
| 1 | 34.1s | 4.98 | 10.02% | 48.97% | 10.92% | 51.69 |
| 2 | 33.9s | 4.60 | 19.81% | 70.94% | 25.54% | 81.08% |
| 5 | 33.9s | 4.15 | 43.01% | 90.62% | 40.31% | 91.46% |
| 10 | 33.8s | 3.78 | 57.39% | 96.15% | 52.85% | 95.69% |

### GPU

| Epoch | Time | Loss | Train Top-1 Accuracy | Train Top-5 Accuracy | Validation Top-1 Accuracy | Validation Top-5 Accuracy |
|---|---|---|---|---|---|---|
| 1 | 6.3s | 4.98 | 9.99% | 48.97% | 10.92% | 52.31% |
| 2 | 5.9s | 4.65 | 18.82% | 69.09% | 23.46% | 79.69% |
| 5 | 5.9s | 4.04 | 43.77% | 91.18% | 46.00% | 93.31% |
| 10 | 6.0s | 3.76 | 57.57% | 96.06% | 53.69% | 96.00% |

**Comment by timmoon10** [comment]
It seems that the discrepancy between CPU and GPU experiments is small relative to network noise and floating-point error. If I repeatedly run experiments, the mean losses on CPU and GPU are typically within a standard deviation of each other. I get similar results with top-1 and top-5 accuracy. Here is a representative plot showing the loss plus/minus a standard deviation:

![loss](https://user-images.githubusercontent.com/4406448/31567650-69afdee4-b025-11e7-85c7-7c6a5521b38f.png)

### Parameters
- AlexNet without image crops, 257 mini-batch size, momentum SGD
- Trained on first 10 classes of ILSVRC2012
- Release build, float data type
- Run on Surface cluster, gpgpu partition: 3 nodes, 2 procs/node
- 10 runs each for CPU and GPU

**Comment by ndryden** [comment]
257 minibatch size?

**Comment by timmoon10** [comment]
I wanted something non-standard.
